Use DictionaryBlocks for Unnest Operator

ajaygeorge opened the pull request:
Use DictionaryBlocks for Unnest Operator

Fixes #13121 

```
== RELEASE NOTES ==

General Changes
* Use DictionaryBlocks for Unnest Operator
```

wenleix commented:
@ajaygeorge can you keep two commits as the original PR? Thanks!

ajaygeorge commented:
Hi @wenleix , Are you referring to these two commits https://github.com/prestosql/presto/pull/901/commits 
https://github.com/prestosql/presto/pull/901/commits/e05502dccd7c917fb7db45ae1793d0d088674938 and https://github.com/prestosql/presto/pull/901/commits/fe9e7c0710e4afbba05366e2f29e1328a3df627c 

This PR brings in changes from three different PRs from prestosql
prestosql/presto#901
prestosql/presto#952
prestosql/presto#1027
So do you want like 2+1+1 = 4 different commits ?

mbasmanova commented:
CC: @oerling

wenleix commented:
@ajaygeorge Correct, please retain the original commits to help review and tracking the codebase.  Thank you !

ajaygeorge commented:
Hi @wenleix . Thanks for the clarification. Since the package structures are different across these two projects `io.prestosql` vs `com.facebook.presto` cherry-picking commits from one repo to another will not suffice.

Is there a recommended workflow for backporting changes from `prestosql` ?

wenleix commented:
@ajaygeorge : I usually just re-implement the change in PrestoDB based on the commit in PrestoSQL.  This also helps me understand the change better.

ajaygeorge commented:
Hi @wenleix , Can you please take a look now. I have added 4 separate commits mirroring the original PRs

wenleix commented on presto-main/src/test/java/com/facebook/presto/block/ColumnarTestUtils.java:
```diff
@@ -13,33 +13,24 @@
 import static org.testng.Assert.assertTrue;
 
-final class ColumnarTestUtils
+public final class ColumnarTestUtils
```
The change here is not shown in https://github.com/prestosql/presto/pull/901/commits/e05502dccd7c917fb7db45ae1793d0d088674938 . Is that done in other commits in PrestoSQL repo? Can you extract that commit as well? :)

ajaygeorge replied:
Removed the other changes coming in from that source file and retained just the single line change as in the commit

yingsu00 commented:
> Hi @wenleix . Thanks for the clarification. Since the package structures are different across these two projects `io.prestosql` vs `com.facebook.presto` cherry-picking commits from one repo to another will not suffice.
> 
> Is there a recommended workflow for backporting changes from `prestosql` ?

@ajaygeorge You can try to use `git format-patch` and `git am` to backport the commits from prestosql. That way it will preserve the author information and commit message. Your current way lost such information. 
cc @wenleix

ajaygeorge commented:
@yingsu00 How is format-patch different from cherry-picking that commit. format-patch and am seems like a 2 step process for cherry-picking.
I have used Github's [multiple author](https://github.blog/2018-01-29-commit-together-with-co-authors/) to attribute commits to the original author.
<img width="985" alt="Screen Shot 2019-09-05 at 4 15 08 PM" src="https://user-images.githubusercontent.com/236188/64389725-a9419f80-cff8-11e9-8edb-a5eb33e82348.png">

wenleix commented:
@yingsu00 The current backport looks good. We can use `git format-patch` in the future but I don't think we need to re-backport it this time.

yingsu00 commented:
> @yingsu00 How is format-patch different from cherry-picking that commit. format-patch and am seems like a 2 step process for cherry-picking.
> I have used Github's [multiple author](https://github.blog/2018-01-29-commit-together-with-co-authors/) to attribute commits to the original author.
> <img alt="Screen Shot 2019-09-05 at 4 15 08 PM" width="985" src="https://user-images.githubusercontent.com/236188/64389725-a9419f80-cff8-11e9-8edb-a5eb33e82348.png">
> 
> Let me know if this looks good

Oh sorry I missed the co-author part. If the author information is preserved it's fine.

yingsu00 commented:
> "Adjust getOffset output in Columnar classes for trimmed element blocks " Also looks good.
> 
> @yingsu00 can you also approve it if it looks good to you ? :)

Yes I'll read it carefully tonight. :)

aweisberg commented:
These commits do not follow our [commit guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines). The **original author field must be preserved**. You also don't need to add the author as a co-author to the commit message.

The commits also don't link to the PR that originally contained the commits.

@wenleix @yingsu00 @ajaygeorge please take a look at the guidelines when you have a chance.

ajaygeorge commented:
@aweisberg 
Retained the original committer using 
`git commit --amend --author=`

Reworked the commit messages to drop original author as `Co-Author` and added `Backport of` which `prestosql` PR

Please take a look and let me know if this looks good.

yingsu00 commented:
@ajaygeorge I see two potential minor issues in the commit messages:
1) The commit guideline says we need to separate subject from body with a blank line. Your commit messages doesn't have this blank line
2) The guideline uses "cherry-pick" while you used "Backport" and there is no absolute URL as suggested: Example:

```
Fix OOM caused by foo in bar

Foo was pack ratting ByteBuffers causing OOM.

Cherry-pick of https://github.com/foo/bar/pull/123 (https://github.com/foo/bar/pull/123)

Co-authored-by: Foo Bar <foo@bar.com>
```
@aweisberg Are we strict on the format of the "Cherry-pick" line?

ajaygeorge commented:
@yingsu00 I modelled the message on https://github.com/prestodb/presto/pull/13169/commits/cad47033f9c62d64317d44f9a070a3917e3b682c. Apart from the commit messages, do you see any other issues with this backport.
@aweisberg Please let me know if the commit messages need further modification.

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ReplicatedBlockBuilder.java:
```diff
@@ -0,0 +1,77 @@
+        checkArgument(count >= 0, "count should be >= 0");
+
+        if (positionCount + count > ids.length) {
+            // Grow capacity
```
The grow capacity can use Arrays.ensureCapacity(). Maybe we can add a method in Arrays that takes requiredCapacity and newCapacity and use it here. But this can happen later. cc @wenleix @mbasmanova

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -0,0 +1,265 @@
+    private void appendId(int sourcePosition)
+    {
+        if (positionCount == ids.length) {
+            // grow capacity
```
Similarly this can use Arrays.ensureCapacity. But this can happen later.

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -0,0 +1,306 @@
+    private int processCurrentPosition()
+    {
+        // Determine number of output rows for this input position
+        int maxEntries = getCurrentMaxEntries();
```
Nit: This function only contains one line and it's only used here. We can inline it.

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -0,0 +1,44 @@
+    static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+
+    // Copied from io.prestosql.spi.block.BlockUtil#calculateNewArraySize
+    static int calculateNewArraySize(int currentSize)
```
Maybe we should move /copy
BlockUtil#calculateNewArraySize to Arrays in presto-array. cc @mbasmanova

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java:
```diff
@@ -0,0 +1,424 @@
+        private String nestedTypeTwo = "NONE";
+
+        @Param({"0.0", "0.2"})
+        private double primitiveNullsRatioNestedOne;  // % of nulls in input primitive elements
```
Nit: Add @SuppressWarnings("unused")

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java:
```diff
@@ -0,0 +1,424 @@
+        private double primitiveNullsRatioNestedOne;  // % of nulls in input primitive elements
+
+        @Param({"0.0", "0.05"})
+        private double rowNullsRatioNestedOne;    // % of nulls in row type elements
```
Nit: Add @SuppressWarnings("unused")

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java:
```diff
@@ -0,0 +1,424 @@
+    }
+
+    @Benchmark
+    public List<Page> unnest(BenchmarkContext context)
```
Nit: Most benchmarks in this repo would put the benchmark at the beginning. I'd recommend put this function before the BenchmarkContext.

yingsu00 commented on presto-main/src/test/java/com/facebook/presto/operator/BenchmarkUnnestOperator.java:
```diff
@@ -0,0 +1,424 @@
+    public static class BenchmarkContext
+    {
+        @Param("varchar")
+        private String replicateType = "varchar";
```
There is only one replicated channel in the benchmark. This is not the reality and missing some fairly important measurements. I would suggest adding multiple replicated channels with different types case in separate commit later.

yingsu00 commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarArray.java:
```diff
@@ -133,12 +132,12 @@ public boolean isNull(int position)
     public int getLength(int position)
     {
-        return getOffset(position + 1) - getOffset(position);
+        return (offsets[position + 1 + offsetsOffset] - offsets[position + offsetsOffset]);
```
Nit: remove the ()

yingsu00 commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarArray.java:
```diff
@@ -133,12 +132,12 @@ public boolean isNull(int position)
+    public int getOffset(int position)
     {
-        return offsets[position + offsetsOffset];
+        return (offsets[position + offsetsOffset] - offsets[offsetsOffset]);
```
Nit: remove the ()

yingsu00 commented on presto-spi/src/main/java/com/facebook/presto/spi/block/ColumnarMap.java:
```diff
@@ -139,9 +138,9 @@ public int getEntryCount(int position)
+    public int getOffset(int position)
     {
-        return offsets[position + offsetsOffset];
+        return (offsets[position + offsetsOffset] - offsets[offsetsOffset]);
```
Nit: remove the ()

wenleix commented:
Merged #13361. Thanks for the contribution!
